pkg/traversal/traversal.go

```diff
@@ -0,0 +1,364 @@
+// Copyright 2020 The Swarm Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package traversal
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+
+	"github.com/ethersphere/bee/pkg/collection/entry"
+	"github.com/ethersphere/bee/pkg/file"
+	"github.com/ethersphere/bee/pkg/file/joiner"
+	"github.com/ethersphere/bee/pkg/manifest"
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/swarm"
+)
+
+var (
+	// ErrInvalidType is returned when the reference was not expected type.
+	ErrInvalidType = errors.New("traversal: invalid type")
+)
+
+// Service is the service to find dependent chunks for an address.
+type Service interface {
+	// TraverseChunkAddresses iterates through each address related to
+	// the supplied one, if possible.
+	TraverseChunkAddresses(context.Context, swarm.Address, swarm.AddressIterFunc) error
```

Comment:
TraverseChunkAddresses name may relate to a single chunk traversal, even if it does not make sense to have that. I would consider naming it without a Chunk word, but have no nice example. Maybe just TraverseAddresses?